### ISSUE TITLE
type test fails to infer correct class in else block

### ISSUE THREAD

**Issue by stevenh** [python/mypy#16771]
**Bug Report**

`if` test using `type` on class1 | class2 doesn't determine the correct class in the else case resulting in errors which are incorrect.

This may already have been reported as it was impossible to search for given the generic nature involving `type` method.

**To Reproduce**

```python
class Duration(int):
    def nanoseconds(self) -> int:
        return int(self)


def output(val: Duration | int) -> None:
    if type(val) is int:
        print(f"type: {type(val)} = int: {val}")
    else:
        print(f"type: {type(val)} = duration: {val.nanoseconds()}")


output(Duration(1))
output(1)
```
[Playground link](https://mypy-play.net/?mypy=latest&python=3.12&gist=ce10afc1b519577034179f3637d2094d)

**Expected Behavior**

No errors

**Actual Behavior**

```
main.py:10: error: Item "int" of "Duration | int" has no attribute "nanoseconds"  [union-attr]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.8.0
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: none

**Comment by stevenh** [comment]
Workaround is to change `else:` -> `elif type(val) is Duration:` but this should be unnecessary.

**Comment by stevenh** [comment]
Using `isinstance` instead of a `type` check doesn't report an issue but in this case where `int` is a parent class of `Duration` that test is incorrect.
```python
def output(val: Duration | int) -> None:
    if isinstance(val, int): # Invalid due to Duration being a sub class of int
        print(f"type: {type(val)} = int: {val}")
    else:
        print(f"type: {type(val)} = duration: {val.nanoseconds()}") # Never called
```

**Comment by erictraut** [comment]
Mypy's behavior is correct here. The type cannot be safely narrowed to just `Duration` in the "else" case. It's possible for other subtypes of `int` to be passed to the function, and they would not pass the test `type(val) is int`.

```python
class OtherIntSubclass(int):
    pass

output(OtherIntSubclass(1))
```

You've already identified the correct way to fix this bug in your code — by using `isinstance`.

**Comment by stevenh** [comment]
Thanks for clarification, coming from other languages my expectation was for exact type matching requirement from hint and hence `Duration | int` meant only `Duration` or `int` exactly, so  `OtherIntSubclass` was not valid.

I believe what you're saying is this is not the case and instead `Duration` or `int` means those classes or any subclasses, is that correct @erictraut?

**Comment by stevenh** [comment]
Answering my own question from [PEP 484](https://peps.python.org/pep-0484/#type-definition-syntax)
> Expressions whose type is a subtype of a specific argument type are also accepted for that argument.

Capturing for others the fix for this case is to invert the test an look for the more specific type, which is easy in this case given one is base type, not so easy for other types as you need know the types of the entire chain.
```
def output(val: Duration | int) -> None:
    if isinstance(val, Duration):
        print(f"type: {type(val)} = duration: {val.nanoseconds()}")
    else:
        print(f"type: {type(val)} = int: {val}")
```

Thanks for the quick response, closing as it is working as intended.
